### PR TITLE
Batches to zarr

### PR DESCRIPTION
**Pull request by leifdenby**
Add `BatchGenerator.to_zarr` and `BatchGenerator.from_zarr` to make it possible to save generated batches to zarr and later load them from zarr. By chunking along the batch dimension this enables fast data-loading at training time.

### CODE REVIEW

**Comment by jhamman** [comment]
@leifdenby - thanks for opening this PR and apologies the review wasn't picked up sooner. 

My main question/comment is about whether or not we want to think about serializing some of the batch-generator's attributes in the Zarr dataset. It seems like without too much effort, we could effectively reconstruct the full BatchGenerator attribute namespace.

Also, as an aside, this fits nicely within the caching-api's element in the xbatcher roadmap: https://xbatcher.readthedocs.io/en/latest/roadmap.html#caching-apis. We hadn't gotten there yet but I'm glad to see this moving.

**Comment by jhamman** [comment]
> Looking at [BatchGenerator.__init__](https://github.com/pangeo-data/xbatcher/blob/dd4108cf1e2ecac9346fe9a426fa74080e8ddae6/xbatcher/generators.py#L100) maybe I should start with input_dims, input_overlap, batch_dims and concat_input_dims

Yes, this is exactly what I was thinking. 

Also, looking at the code coverage report, it looks like we're in pretty good shape but could use a bit more testing on the edge cases. I'll leave a few more comments in the code to highlight areas that could use a test.

File: xbatcher/generators.py [hunk 1]
```diff
@@ -144,3 +144,54 @@ def _iterate_batch_dims(self, ds):
...
+        to the available RAM on the computere where you are doing ML model
+        training
+        """
+        batch_datasets = list(self)
+        # can't call the batch dimension `batch` because Dataset.batch is used
+        # for the batch acccessor. Instead we'll call it `batch_number`
+        ds_all = xr.concat(batch_datasets, dim='batch_number').reset_index(
+            'sample'
+        )
+        if 'batch' in chunks:
```
Comment (jhamman) [inline xbatcher/generators.py:161]: test when `'batch' not in chunks`

File: xbatcher/generators.py [hunk 2]
```diff
@@ -144,3 +144,54 @@ def _iterate_batch_dims(self, ds):
...
+            chunks['batch_number'] = chunks.pop('batch')
+
+        if len(chunks) > 0:
```
Comment (jhamman) [inline xbatcher/generators.py:164]: test when `len(chunks) == 0`

File: xbatcher/generators.py [hunk 3]
```diff
@@ -144,3 +144,54 @@ def _iterate_batch_dims(self, ds):
...
+    the batches from a zarr store that was previously created with
+    `BatchGenerator.to_zarr`
+    """
+
+    def __init__(self, path):
+        self.ds_batches = xr.open_zarr(path)
+        self.path = path
+
+    def __iter__(self):
+        for batch_id in self.ds_batches.batch_number.values:
```
Comment (jhamman) [inline xbatcher/generators.py:188]: I'm not exactly why but codecov think something in this for loop is not being covered by the existing tests. Perhaps its the empty iterable (`.values) or it could be the `if` statement in line 194. Any thoughts?

**Review by jhamman — COMMENTED** [review]
a few pointers for possible test improvements

**Review by RichardScottOZ — COMMENTED** [review]
typo in computere by the way
